### ISSUE TITLE
python 3.6 support

### ISSUE THREAD

**Issue by lestatcheb** [stripe/stripe-python#329]
stripe client doesn't work with python 3.6 (tried last stripe==1.59.0)

```
  File "/mnt/storage/code/venv/lib/python3.6/site-packages/stripe/resource.py", line 466, in create
    response, api_key = requestor.request('post', url, params, headers)
  File "/mnt/storage/code/venv/lib/python3.6/site-packages/stripe/api_requestor.py", line 150, in request
    method.lower(), url, params, headers)
  File "/mnt/storage/code/venv/lib/python3.6/site-packages/stripe/api_requestor.py", line 339, in request_raw
    util.log_debug('API response body', body=rbody)
  File "/mnt/storage/code/venv/lib/python3.6/site-packages/stripe/util.py", line 86, in log_debug
    msg = logfmt(dict(message=message, **params))
  File "/mnt/storage/code/venv/lib/python3.6/site-packages/stripe/util.py", line 129, in logfmt
    return ' '.join([fmt(key, val) for key, val in sorted(props.items())])
  File "/mnt/storage/code/venv/lib/python3.6/site-packages/stripe/util.py", line 129, in <listcomp>
    return ' '.join([fmt(key, val) for key, val in sorted(props.items())])
  File "/mnt/storage/code/venv/lib/python3.6/site-packages/stripe/util.py", line 122, in fmt
    val = str(val)
BytesWarning: str() on a bytes instance
```

**Comment by ob-stripe** [comment]
@lestatcheb That's weird, the library is tested against 3.6: https://travis-ci.org/stripe/stripe-python/jobs/244747027.

Can you share the code that led to this exception? Make sure to redact your API keys.

**Comment by lestatcheb** [comment]
Hi, sure.

```
> import stripe
> stripe.api_key = 'my_api_key'
> stripe.api_version = '2017-02-14'
> token_create_data = {
>     'number': '4242424242424242',
>     'name': 'TEST TESTOV',
>     'exp_month': '12',
>     'exp_year': '2019',
>     'cvc': '123',
> }
> res = stripe.Token.create(card=token_create_data)


message='Request to Stripe api' method=post path=https://api.stripe.com/v1/tokens
message='Stripe API response' path=https://api.stripe.com/v1/tokens response_code=200
---------------------------------------------------------------------------
BytesWarning                              Traceback (most recent call last)
<ipython-input-1-f8c858aee1e6> in <module>()
     11     'cvc': '123',
     12 }
---> 13 res = stripe.Token.create(card=token_create_data)

/mnt/storage/code/venv/lib/python3.6/site-packages/stripe/resource.py in create(cls, api_key, idempotency_key, stripe_account, **params)
    464         url = cls.class_url()
    465         headers = populate_headers(idempotency_key)
--> 466         response, api_key = requestor.request('post', url, params, headers)
    467         return convert_to_stripe_object(response, api_key, stripe_account)
    468

/mnt/storage/code/venv/lib/python3.6/site-packages/stripe/api_requestor.py in request(self, method, url, params, headers)
    148     def request(self, method, url, params=None, headers=None):
    149         rbody, rcode, rheaders, my_api_key = self.request_raw(
--> 150             method.lower(), url, params, headers)
    151         resp = self.interpret_response(rbody, rcode, rheaders)
    152         return resp, my_api_key

/mnt/storage/code/venv/lib/python3.6/site-packages/stripe/api_requestor.py in request_raw(self, method, url, params, supplied_headers)
    337         util.log_info(
    338             'Stripe API response', path=abs_url, response_code=rcode)
--> 339         util.log_debug('API response body', body=rbody)
    340         if 'Request-Id' in rheaders:
    341             util.log_debug('Dashboard link for request',

/mnt/storage/code/venv/lib/python3.6/site-packages/stripe/util.py in log_debug(message, **params)
     84
     85 def log_debug(message, **params):
---> 86     msg = logfmt(dict(message=message, **params))
     87     if _console_log_level() == 'debug':
     88         print(msg, file=sys.stderr)

/mnt/storage/code/venv/lib/python3.6/site-packages/stripe/util.py in logfmt(props)
    127             key = repr(key)
    128         return '{key}={val}'.format(key=key, val=val)
--> 129     return ' '.join([fmt(key, val) for key, val in sorted(props.items())])
    130
    131

/mnt/storage/code/venv/lib/python3.6/site-packages/stripe/util.py in <listcomp>(.0)
    127             key = repr(key)
    128         return '{key}={val}'.format(key=key, val=val)
--> 129     return ' '.join([fmt(key, val) for key, val in sorted(props.items())])
    130
    131

/mnt/storage/code/venv/lib/python3.6/site-packages/stripe/util.py in fmt(key, val)
    120         # translated incorrectly.
    121         if not isinstance(val, str):
--> 122             val = str(val)
    123         if re.search(r'\s', val):
    124             val = repr(val)

BytesWarning: str() on a bytes instance
```

```
> import stripe
> stripe.VERSION
'1.59.0'
```

```
> python --version
Python 3.6.1
> uname -a
Linux x250 4.11.6-1-ARCH #1 SMP PREEMPT Sat Jun 17 08:19:42 CEST 2017 x86_64 GNU/Linux
```

**Comment by lestatcheb** [comment]
Also we use python with -bb argument:

```
> python --help
...
-b     : issue warnings about str(bytes_instance), str(bytearray_instance)
         and comparing bytes/bytearray with str. (-bb: issue errors)
...
```

**Comment by ob-stripe** [comment]
@lestatcheb Thanks for the additional information! I was able to reproduce the issue trivially with your example.

To be fair, Python issues a `BytesWarning`, not an error, but since you use the `-bb` flag warnings are turned into errors. That said, I agree that this needs to be fixed anyway.

The problem originates here: https://github.com/stripe/stripe-python/blob/e0fe78181af1b6dfcbf92414385a8dfd7de447f9/stripe/api_requestor.py#L339. With Py3, `rbody` will be a `bytes` instance which the log functions don't handle properly.

A simple fix would be to add this at the beginning of the `fmt` method:

```py
        if sys.version_info >= (3, 0) and isinstance(val, bytes):
            val = val.decode('utf-8')
```

but there might be a better way.

cc @deontologician-stripe since you wrote the logging feature -- wdyt?

**Comment by jleclanche** [comment]
I'd like to suggest adding `PYTHONWARNINGS = all` in [tox setenv](https://github.com/stripe/stripe-python/blob/master/tox.ini#L18)

**Comment by ob-stripe** [comment]
Oh wow, nice catch! It should of course be `-W all`.

**Comment by brandur-stripe** [comment]
Thanks for reporting! @ob-stripe patched the problem and we've released a fix in 1.71.1.
